sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala

```diff
@@ -418,12 +430,21 @@ object GpuShuffleExchangeExecBase {
               // Get a non-empty batch or the last batch. So still need to
               // check if it is empty for the later case.
               if (batch.numRows > 0) {
-                partitioned = getParts(batch).asInstanceOf[Array[(ColumnarBatch, Int)]]
-                partitioned.foreach(batches => {
-                  metrics(GpuMetric.NUM_OUTPUT_ROWS) += batches._1.numRows()
-                })
-                metrics(GpuMetric.NUM_OUTPUT_BATCHES) += partitioned.length
-                at = 0
+                val spillableBatch = SpillableColumnarBatch(batch,
+                    ACTIVE_ON_DECK_PRIORITY)
+                partitionedIter = withRetry(spillableBatch,
+                    splitSpillableInHalfByRows) { spillable =>
+                  val cb = spillable.getColumnarBatch()
+                  getParts(cb).asInstanceOf[Array[(ColumnarBatch, Int)]]
+                }
+                if (partitionedIter.hasNext) {
```

Comment:
we should generalize this branch (where partitionedIter.hasNext is true). we do it here and also on line 417